### ISSUE TITLE
the nebula question

### ISSUE THREAD

**Issue by codefromthecrypt** [micrometer-metrics/micrometer#1787]
While I've worked in micrometer before, I'm kindof looking deeper than usual. One thing that kindof popped up was that build tasks are built above Netflix Nebula, along with a couple custom gradle plugins. I'm aware that in Gradle, custom plugins is fairly routine so that's not necessarily a bad thing.. Just curious if Nebula is a conscious intent, or a relic.

It seems helpful to work with close parties who have similarly large and diverse codebases and answers to common questions. For example, we're getting some issues in the last few days about citations etc. I've some history working in the Armeria project (ping contributor @trustin) and have found their way to address dependencies kindof straightforward, even if it is still gradle https://github.com/line/armeria/blob/master/dependencies.yml

So, basically I'm opening this up as a kindof way to get a read on what folks want to do with the build moving forward, if gradle is still desirable, and if that's the case, if nebula is.. and if that's the case, what's the strategy with movement over there and what we're doing here.

**Comment by anuraaga** [comment]
Almost all of Nebula's features have been folded into Gradle by now so it's probably reasonable to consider dropping it.

I use a pattern of defining a dependencies project and using Gradle's `java-platform` for managing dependencies.

https://github.com/curioswitch/curiostack/blob/master/tools/curiostack-bom/build.gradle.kts

I use static versions and Ben's `com.github.ben-manes.versions` to find new versions before releasing. Currently, once-a-day I incur a 15+ min overhead (maybe since I'm in Japan, dunno why it's so slow) refreshing the Gradle project for micrometer due to the dynamic versions, it would definitely be great to use static version numbers.

One gotcha is making sure Maven publishes resolved versions, it doesn't seem so hard but actually I haven't tried this feature yet, my plugin manually rewrites the POM since I hadn't found it yet!
https://docs.gradle.org/current/userguide/publishing_maven.html#publishing_maven:resolved_dependencies

I'd be happy to make a change to try this pattern if there's interest (would first just change the dependency management without removing nebula itself which has other features too)

**Comment by adriancole** [comment]
non-binding, I'd be in favor of using base functionality that now exists.

I say non-binding as I'm drive-by guy at the moment, and tommy has been
doing the releases so would have highest impact to any build change

**Comment by shakuzen** [comment]
If parts of the build contain functionality by non-official Gradle plugins that are now natively supported by Gradle itself or one of its official plugins, I'm generally in favor of moving in that direction. I simply have not kept up with following all of Gradle's evolution so I'm likely not aware of what these are or how to configure them without learning the new thing Gradle has made.

That said, there has been some talk of dependency versions in this issue. Just so we're on the same page, the versions in the project are using Gradle version ranges and on branches other than `master` (maintenance branches like `1.1.x` and `1.3.x`) are in addition using Gradle's built-in dependency locking feature. There's room for improvement here, but I think it's separate from anything Nebula related unless I'm misunderstanding.

**Comment by anuraaga** [comment]
Thanks for the clarification, just assuming it's nebula is my mistake. Indeed then it's a separate issue of whether to change the dynamic version system (realized that probably means that my slowness is only on master, not other branches)

**Comment by anuraaga** [comment]
Looked at the current build a bit more, and to clarify I guess technically we aren't applying the nebula plugin, we apply `spring-release-plugin` which itself applies nebula.

https://github.com/spring-gradle-plugins/spring-release-plugin

These features seem very important for maintaining and publishing this repo
- Configures license header checks on files. Also places a default license header template
in `gradle/licenseHeader.txt` which is applied to any file needing a license header. Conditions
release builds upon having proper license headers in all source files.
- Configures Asciidoc compilation and uploading to Github pages.
- Adds the `nebula.release` plugin.
- Configures `nebula.bintray-publishing`, automatically filling all the relevant Bintray configuration
based on any git remote attached to your repository that has an organization containing the word `spring`.

These features seem of questionable usefulness
- Adds a properties section to a generated POM publication with information about the
environment and status of the repository when the artifact was built.
  - Would say a properties file in the jar is a better location than POM for this
- Adds SCM information to the POM.
  - Ditto
- Applies the Nebula contacts plugin, requiring information about chief committers.
  - Not sure what this provides over the `developers` section of the Maven POM

These are all provided by nebula right now but are natively supported by Gradle and ideally could be replaced with native functionality

- Configures a `maven-publish` publication.
- Substitutes dynamic revisions in the generated POM with fixed versions.
- Adds exclude information to the POM.
- Generates a javadoc jar
- Generates a source jar

Unfortunately, it's all or nothing as applying `io.spring.publishing` applies both useful and deprecated features. 

So if we wanted to clean something up here, options are

- Update `spring-release-plugin` to replace nebula usage with native where it can
- Fold the important features of `spring-release-plugin` into this build file and stop using it. It's a fairly small amount of code which will be even smaller if written in a build file instead of a Java plugin. 
 https://github.com/spring-gradle-plugins/spring-release-plugin/tree/master/src/main/groovy/io/spring/gradle/release

For reference, I found one other project using `spring-release-plugin` which doesn't look like a spring project in fact - https://github.com/search?q=%22io.spring.publishing%22&type=Code There are three different `micrometer` projects using it though, I wonder if they should all be merged into this monorepo?

**Comment by shakuzen** [comment]
> * Configures Asciidoc compilation and uploading to Github pages.

This isn't needed for this repo. We have a separate docs repo that deploys a static website: https://github.com/micrometer-metrics/micrometer-docs/

> Adds a properties section to a generated POM publication with information about the
> environment and status of the repository when the artifact was built.
> 
>  * Would say a properties file in the jar is a better location than POM for this

Agreed. In fact, all of the properties do end up in the JAR's META-INF in `MANIFEST.MF` and a properties file already. Not having them in the POM would be a welcome change.

> * Adds SCM information to the POM.
>   
>   * Ditto
> 
> * Applies the Nebula contacts plugin, requiring information about chief committers.
>   
>   * Not sure what this provides over the `developers` section of the Maven POM

This is what is used for the `developers` section of the POM. It might also be used as one of those previously mentioned metadata added to the manifest and a properties file in the JAR. I'm not sure if it is used as metadata anywhere else like Bintray, or if Bintray extracts that info from the `developers` section of the POM.

SCM info in the POM as well as the `developers` section are requirements for artifacts published to Maven Central: https://central.sonatype.org/pages/requirements.html

> There are three different `micrometer` projects using it though, I wonder if they should all be merged into this monorepo?

I'd like those other projects to be able to release at their own cadence and therefore not be in the main repository.

> So if we wanted to clean something up here, options are

Either option could work. As mentioned, there are a couple other projects that changes would need to be propagated across, which I guess is the appeal of having a plugin and being able to bump the version rather than copy-and-pasting to each project. On the other hand, the plugins haven't been updated in a while, and I've never released any of them, so there is a bit of a learning curve there.

**Comment by anuraaga** [comment]
> This is what is used for the developers section of the POM.

Ah thanks for pointing this out. I should mention that Gradle has native methods for setting developers so probably not much code difference vs contacts plugin. I noticed contacts plugins adds a few extra `properties` to the POM.

For SCM, I forgot about the standard links to GitHub. Personally, I'd be more comfortable if those are explicitly defined statically, using Gradle's native methods, than automatically based on what the repo's remotes happen to be set to at the time.

For reference, here's what the native config looks like

https://github.com/curioswitch/curiostack/blob/master/build.gradle#L82
